src/charm.py

```diff
@@ -165,14 +165,18 @@ def _on_config_changed(self, event: ConfigChangedEvent) -> None:
             )
             event.defer()
             return
-        if self._certificates_relation_created:
+        if not self._csr_is_stored:
+            self._revoke_existing_certificates()
+            self._generate_csr()
+            self.unit.status = ActiveStatus()
+            return
+        elif self._certificates_relation_created:
             self._revoke_existing_certificates()
             self._generate_csr()
             self._request_certificate()
             self.unit.status = WaitingStatus("Waiting for certificate to be available")
         else:
-            self._generate_csr()
-            self.unit.status = ActiveStatus()
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created.")
```

Comment:
I'm unsure about the status handling here, we set to active if the csr is not stored but to waiting if it is stored...

Reply:
I'm not sure about the `ActiveStatus` either. The charm can't really be in active state until the cert is there, right?
How about checking whether the `certificates` relation is created first?
```python
if not self._certificates_relation_created:
    self.unit.status = BlockedStatus("Waiting for certificate to be available")
    event.defer()
    return
self._revoke_existing_certificates()
self._generate_csr()
self._request_certificate()
self.unit.status = WaitingStatus("Waiting for certificate to be available")
```
And then `_on_certificate_available` would set the ActiveStatus.

Reply:
How about this flow ?
```
        if not self._certificates_relation_created:
            self.unit.status = BlockedStatus("Waiting for certificate relation to be created")
            event.defer()
            return

        self._revoke_existing_certificates()
        self._generate_csr()
        if not self._csr_is_stored:
            self._generate_csr()
            self.unit.status = WaitingStatus("Waiting csr to be stored.")
            event.defer()
            return
        self._request_certificate()
        self.unit.status = WaitingStatus("Waiting for certificate to be available")
        ```

Reply:
I don't think we need this additional if, because `_csr_is_stored` already does:
```python
if self._csr_is_stored:
    csr_secret = self.model.get_secret(label=CSR_SECRET_LABEL)
    csr_secret.set_content(content=csr_secret_content)
else:
    self.app.add_secret(content=csr_secret_content, label=CSR_SECRET_LABEL)
```